File: DSA-In-C&C++/03-DSA-Problems/03-Arrays/2-Medium/01-TwoSum.cpp
/*
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
*/

#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target){
    vector<int> ans;
    for(int i = 0; i < nums.size(); i++){
        for(int j = i + 1; j < nums.size(); j++){
           if(nums[i] + nums[j] == target){
            ans.push_back(i);
            ans.push_back(j);
            return ans;
           }
        }
    }
}

/*
An unordered map is a data structure commonly found in programming languages, such as C++, and it functions as a key-value store. 
It allows for efficient insertion, deletion, and lookup operations with an average time complexity of 
O(1). The "unordered" aspect means that the elements in the map are not stored in any specific order; instead, they are organized internally using a hash table.
*/

vector<int> twoSum2(vector<int>& nums, int target){
    vector<int> ans;
    unordered_map<int, int> myMap;
    for(int i = 0; i < nums.size(); i++){
        if(myMap.find(nums[i]) != myMap.end()){
            ans.push_back(myMap[nums[i]]);
            ans.push_back(i);

            return ans;
        }

        int remaining = target - nums[i];
        myMap[remaining] = i;
    }
}


int main(){
    vector<int> nums = {2, 7, 11, 15};
    int target = 18;

    vector<int> ans =  twoSum2(nums, target);

    for(int it:ans){
        cout<<it<<", ";
    }
    return 0;
}

File: DSA-In-JS&TS/DSA-Problems/1-Array/2-Medium/01-TwoSum.js
/*
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
*/

//! O(n^2)
// var twoSum = function(nums, target) {
//     for(let i = 0; i < nums.length; i++){
//         for(let j = i + 1; j < nums.length; j++){
//             if(nums[i] + nums[j] === target){
//                 return [i, j]
//             }
//         }
//     }
//     return []
// };

// console.log(twoSum([2, 3, 5, 1], 4))

/*
In JavaScript, a Map is a built-in object that stores key-value pairs. Unlike plain objects, 
where keys are always treated as strings or symbols, a Map allows keys of any data type, including objects, functions, and primitive types.
*/

//! O(n)
var twoSum = function(nums, target) {
    const myMap = new Map()
    for(let i = 0; i < nums.length; i++){
        if(myMap.has(nums[i])){
            return [myMap.get(nums[i]), i]
        }
        const remaining = target - nums[i]
        
        myMap.set(remaining, i);

    }
    return []
};

console.log(twoSum([2, 3, 5, 1], 4))
